src/cdi_ndi/NDI_TN.cc

```diff
@@ -0,0 +1,279 @@
+//----------------------------------*-C++-*-----------------------------------//
+/*!
+ * \file   cdi_ndi/NDI_TN.cc
+ * \author Ben R. Ryan
+ * \date   2020 Feb 4
+ * \brief  NDI_TN member definitions.
+ * \note   Copyright (C) 2019-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//----------------------------------------------------------------------------//
+
+#include "NDI_TN.hh"
+#include <cmath>
+
+namespace rtt_cdi_ndi {
+
+//----------------------------------------------------------------------------//
+// CONSTRUCTORS
+//----------------------------------------------------------------------------//
+/*!
+ * \brief Constructor for NDI reader specific to TN reaction data.
+ *
+ * This constructor opens an NDI file, navigates to the appropriate data, reads
+ * the data into internal buffers, and closes the file. For more details on NDI,
+ * see https://xweb.lanl.gov/projects/data/nuclear/ndi/ndi.html
+ *
+ * \param[in] gendir_in path to gendir file
+ * \param[in] library_in name of requested NDI data library
+ * \param[in] reaction_in name of requested reaction
+ * \param[in] discretization_in type of energy discretization to request
+ */
+NDI_TN::NDI_TN(const std::string &gendir_in, const std::string &library_in,
+               const std::string &reaction_in,
+               const DISCRETIZATION discretization_in)
+    : NDI_Base(gendir_in, "tn", library_in, reaction_in, discretization_in),
+      control(1234567) {
+  int gendir_handle, dataset_handle, ndi_size_error;
+  constexpr int c_str_len = 4096;
+  char c_str_buf[c_str_len];
+
+  control.initialize(1, rng);
+
+  // Open gendir file (index of a complete NDI dataset)
+  SAFE_NDI(NDI2_open_gendir(&gendir_handle, gendir.c_str()));
```

Comment:
- I will eliminate the macro in favor of in-place checking.
- I still prefer to have the NDI library handle the checking for the existence of the `gendir` file, because it controls the actual file. Is that OK?
- Added length checks to strings passed to constructor in `NDI_Base.hh`
- Added checks for default values in `gendir_handle` and `dataset_handle`
- I worry that preserving `gendir_handle` and `dataset_handle` could be misleading here (I'm not familiar with RAII so I may have misunderstood) because after we call `NDI2_close_gendir(gendir_handle)` at the end of the `NDI_TN` constructor, these handles have no meaning